### PR TITLE
#160 Use portable resource URL for Custom CSS URL

### PR DESCRIPTION
**Pull request by adanielvv**
Before (I manually changed the url to remove the warning):
screenshot before

After:
screenshot after

See #160

### CODE REVIEW

**Comment by seamuslee001** [comment]
I'm a -1 on this at the moment, not everyone will necessarily be using the [civicrm.files]/ext as the location for Extensions. The current code makes sense as it will always resolve to the correct location no matter where the extensions are. I think what we may want is if it is in the files/ext we replace it to be that otherwise we don't alter it.

**Comment by DanielvV** [comment]
@seamuslee001 
> not everyone will necessarily be using the [civicrm.files]/ext as the location for Extensions.

"ext" is not in my code proposal.

> The current code makes sense as it will always resolve to the correct location no matter where the extensions are.

My code proposal doesn't change that. :)

First, the absolute url to the extension is requested (original code):
`    $baseUrl = CRM_Extension_System::singleton()->getMapper()->keyToUrl('org.civicrm.shoreditch');`

Then, I get the absolute URL to `[civicrm.files]/` (my code).
`    $filesVarUrl = '[civicrm.files]/';`
`    $filesUrl = Civi::paths()->getUrl($filesVarUrl, 'absolute');`

Lastly, I replace the url part that links to CiviCRM files in `$baseUrl` with the variable.
`    $baseVarUrl= str_replace($filesUrl, $filesVarUrl, $baseUrl);`

> I think what we may want is if it is in the files/ext we replace it to be that otherwise we don't alter it.

If for some reason the extensions dir is not somewhere inside the `[civicrm.files]/` folder, `str_replace` won't alter the url. In every other case only part of the url is replaced with a variable that has the same value as the part replaced.

The url is saved in the database, so for existing installations the url will not be altered on upgrade.
Only the bullet in the settings page will jump (if the extensions are under `[civicrm.files]/*`):
If the user fixed the warning the bullet will jump to the left (see above).
Otherwise the bullet will jump to the right, making it very easy for the user to fix the warning.

**Comment by seamuslee001** [comment]
Sorry Daniel, must have misread the code or something, i thought it was removing the extension base url and got confused with the screenshots, re-looking at the code i am a +1 on this, should we also be considering the other portable options?

**Comment by DanielvV** [comment]
I think this is the only url that is saved in the database.

File: shoreditch.php [hunk 1]
```diff
@@ -144,7 +144,10 @@ function shoreditch_civicrm_buildForm($formName) {
   }
   elseif ($formName == 'CRM_Admin_Form_Setting_Url') {
     $baseUrl = CRM_Extension_System::singleton()->getMapper()->keyToUrl('org.civicrm.shoreditch');
-    $cssUrl = CRM_Utils_File::addTrailingSlash($baseUrl, '/') . 'css/custom-civicrm.css';
+    $filesVarUrl = '[civicrm.files]/';
+    $filesUrl = Civi::paths()->getUrl($filesVarUrl, 'absolute');
+    $baseVarUrl = str_replace($filesUrl, $filesVarUrl, $baseUrl);
+    $cssUrl = CRM_Utils_File::addTrailingSlash($baseVarUrl, '/') . 'css/custom-civicrm.css';
```
Comment (AkA84) [inline shoreditch.php]: Maybe better reshuffle the code like this
```php
$filesUrlPlaceholder = '[civicrm.files]/';

$extUrl = CRM_Extension_System::singleton()->getMapper()->keyToUrl('org.civicrm.shoreditch');
$filesUrl = Civi::paths()->getUrl($filesUrlPlaceholder, 'absolute');
$baseVarUrl = str_replace($filesUrl, $filesUrlPlaceholder, $extUrl);
$legacyCssUrl = CRM_Utils_File::addTrailingSlash($baseVarUrl, '/') . 'css/custom-civicrm.css';

Civi::resources()
  ->addScriptFile('org.civicrm.shoreditch', 'js/urlSettingsForm.js')
  ->addVars('shoreditch', array('cssUrl' => $legacyCssUrl));
```

**Comment by AkA84** [comment]
@adanielvv ping

**Comment by MikeyMJCO** [comment]
What's needed at this point to move this issue along?

File: shoreditch.php (hunk 1 above)
Comment (JoeMurray) [reply shoreditch.php]: @adanielvv this approach seems like an improvement

**Comment by JoeMurray** [comment]
The error message is confusing because there is no option that can be changed to a variable to get rid of the error.

When a variable like [civicrm.files]/ext is used, then from one reasonable perspective, there should not be an error message displayed. So if there is a variable used in Extension Resource URL, then one might think suppressing the error would be reasonable.

However, the underlying issue is that when the site is migrated, eg from staging to production, the shoreditch theme will get confused since it has stored an absolute url in the db that is no longer valid. So it's reasonable for some sort of error to be displayed as there is an issue that site admins need to know about.

Am I right that this PR will not address problem that was intended to be addressed by using variables in resource urls, namely, site portability?

**Comment by AkA84** [comment]
@adanielvv any update on this?

**Comment by mlutfy** [comment]
I think this can be closed now that the CSS is managed by frontend/backend theme settings?

**Comment by jamienovick** [comment]
Agreed, closing!

**Pull request closed without merging**